File: 2 - Sorting and Order Statistics/2.1 - Heap Sort/Code/build_heap.py
"""
build_heap.py

"""

from math import floor
from .heapify import Heapify


class BuildHeap(Heapify):
    """Build Max Heap Class"""

    def build_max_heap(self):
        """Construct a binary max heap"""
        for i in range(floor(self.heap_size / 2), -1, -1):
            self.max_heapify(i)

    def build_min_heap(self):
        """Construct a binary min heap"""
        for i in range(floor(self.heap_size / 2), -1, -1):
            self.min_heapify(i)

File: 2 - Sorting and Order Statistics/2.1 - Heap Sort/Code/BuildHeap.mjs
/**
 * BuildHeap.mjs
 *
 */

import Heapify from "./Heapify.mjs";

class BuildHeap extends Heapify {
    constructor(heap) {
        super(heap);
    }

    buildMaxHeap() {
        for (let i = Math.floor(this.heapSize / 2); i > -1; i--) {
            this.maxHeapify(i);
        }
    }

    buildMinHeap() {
        for (let i = Math.floor(this.heapSize / 2); i > -1; i--) {
            this.minHeapify(i);
        }
    }
}

const main = () => {
    const array = [4, 1, 3, 2, 16, 9, 10, 14, 8, 7];
    console.log(`Input: ${array}`);

    let constructMaxHeap = new BuildHeap(array);
    constructMaxHeap.buildMaxHeap();
    console.log(`Max Heap: ${constructMaxHeap.heap}`);

    let constructMinHeap = new BuildHeap(array);
    constructMinHeap.buildMinHeap();
    console.log(`Min Heap: ${constructMinHeap.heap}`);
};

// main();

export default BuildHeap;
